### ISSUE TITLE
Server-side rendering with Blazor WebAssembly

### ISSUE THREAD

**Issue by mifopen** [dotnet/aspnetcore#24098]
### Is your feature request related to a problem? Please describe.
I have the following requirements for my app:
1. No WebSockets (so no Blazor Server)
1. Server-side rendering (very SEO-sensitive app)
1. I want to use Blazor WebAssembly

I wonder if Blazor can suit them all.

### Describe the solution you'd like
It almost works after the following steps:

 1. Install the hosted version of `blazorwasm` template app.
    ```
    dotnet new blazorwasm --hosted -o HostedWasm
    ```
 1. Copy `_Hosted.cshtml` from `blazorserver` template to the `Pages` folder.
 1. Change `endpoints.MapFallbackToFile("index.html")` to `endpoints.MapFallbackToPage("/_Host")` in `Startup.cs` of the server
 1. Change `<script src="_framework/blazor.server.js"></script>` to `<script src="_framework/blazor.webassembly.js"></script>` in `_Host.cshtml`

Now if you `dotnet run` the server you will be able to navigate template pages with prerendered HTML.

But there is a problem: there is no way to initialize the state of components before the first render.
For example, if I store counter value in the memory cache and save it on each increment, then after page refresh I will see
pre-rendered value from the cache... and then 0 (uninitialized component state)

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
@javiercn I believe we have a separate issue tracking client-side prerendering. Can you please dupe this one to it - couldn't find it.

**Comment by javiercn** [comment]
This is already possible today. The following doc provides an overview of what you need to do to support it. See [here](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/additional-scenarios?view=aspnetcore-3.1#support-prerendering-with-authentication) for details

**Comment by mifopen** [comment]
@javiercn, thanks for the answer, but the doc doesn't mention a solution to my problem. Yes, it is possible to server-side render .razor file and the connect Blazor WebAssembly to it, but you will notice UI flickering from server-side rendered view to uninitialized component and back again. The state that we use on the server to render the page doesn't hydrated to client, so we have uninitialized components on first render

**Comment by javiercn** [comment]
@mifopen we don't have an out of the box solution for preserving the state from the prerendered component. That's up to the app developer for now. I believe there's a separate issue for this request, but I can't find the issue number, so I'll leave this one open for the time being.

**Comment by mkArtakMSFT** [comment]
This is a dupe of #14434

**Comment by srpeirce** [comment]
Maybe these? 
https://github.com/dotnet/aspnetcore/issues/22202
https://github.com/dotnet/aspnetcore/issues/22819
https://github.com/dotnet/aspnetcore/issues/21285

**Comment by hoksource** [comment]
Leaving a message to follow this one. Angular universal was able to do this with years of work. Angular universal does server side rendering for seo sensitive stuffs. I think blazor might be able to catch up. The more users using wasm the higher the chance dom manipulation will be enabled for wasm. C# dependency injection improvements might be on the works also.

**Comment by hoksource** [comment]
One use case should be multiple server side pages /index and /item/{itemId}. the server will render head tags so it will be passing data like item object fetched from db to blazor. And the server should be able to choose which blazor route/component to render like so.

Server rendered
```
<Script src="blazorapp.js"></script>
<Script>
var itemmodel={server-rendered-json};
var chosenComponentRoute="item"

var blazorElementContainer=document.getElementByID("my-blazor-container-div");

blazor.bootsrap/render(blazorElementContainer, chosenComponentRoute, itemModel);

</Script>
<div id="my-blazor-container-div"></div>
```

Routes/component should probably be lazy loaded because blazor is too heavy right now.

**Comment by mifopen** [comment]
@javiercn, could you please explain why the issue is marked as `Done`? I don't see any PRs about it and all suggestions from the thread above do not solve the problem and just masking it with some hacks.

**Comment by mkArtakMSFT** [comment]
@javiercn do we have any doc updates to do as this is now done?

**Comment by javiercn** [comment]
I have not worked out the doc, but it's all in the doc that I wrote for the PR:

@guardrex do you think you can use that as a basis for the content?
https://github.com/dotnet/aspnetcore/pull/25203#issuecomment-680105139

Main points are to call out the scenarios and to call out the limitations, explicitly that the component types as well as the parameters must be defined in an assembly referenced by the Blazor WebAssembly application.

**Comment by guardrex** [comment]
> call out the limitations, explicitly that the component types as well as the parameters must be defined in an assembly referenced by the Blazor WebAssembly application.

I thought we had it all covered, but I'll have to check against your linked comment and see. The coverage thus far is ...

https://docs.microsoft.com/aspnet/core/blazor/components/integrate-components#render-components-from-a-page-or-view
https://docs.microsoft.com/aspnet/core/mvc/views/tag-helpers/built-in/component-tag-helper

**Comment by javiercn** [comment]
@guardrex this is new. This gives you the ability to prerender the app on the server with parameters that get flowed down to the client, similar to SSB and to define the root component from MVC in the same way is done for server-side Blazor.

**Comment by guardrex** [comment]
I think this is tracked by https://github.com/dotnet/AspNetCore.Docs/issues/18431. I understand that the 5.0 approach will be different than what DR and I were discussing on the issue, which was more like 3.x hacks at the time.

It looks like https://github.com/dotnet/AspNetCore.Docs/issues/11366 will be covered by that work, too.

The best scenario for me is to wait on until RC1 drops to tackle this one. In the RC1 period, I'll also be working on ...

* CSS isolation
* Blazor WASM security for Identity 2.0
* ~4 other smaller 5.0 updates

**Comment by javiercn** [comment]
@guardrex sounds fine to me, just want to make sure you have the content available.
